feat(service): assign to a copied study all groups of the user

MartinBelthle opened the pull request.

MartinBelthle committed:
feat(service): assign to a copy study all groups of the user

ghost commented:
### Features

*  **service:** assign to a copy study all groups of the user ([620e76d](https://github.com/AntaresSimulatorTeam/AntaREST/commit/620e76d6ce2eaf27aebd197e16143f1d65d1dcda))
*  **api:** improve documentation and implementation of `copy_study` endpoint ([5d098b3](https://github.com/AntaresSimulatorTeam/AntaREST/commit/5d098b31cad9e141ce4a6f26f9c50013952f61aa))


### Tests

*  **api:** simplify integration tests by using pytest fixtures ([610a52b](https://github.com/AntaresSimulatorTeam/AntaREST/commit/610a52b8ebc6016c50faa5deee43cb1c62e0d4dd))


### Bug Fixes

*  **api:** response comments PR ([e26c114](https://github.com/AntaresSimulatorTeam/AntaREST/commit/e26c114e3e189b45a2dba4dfb9823b4fd8073c9d))
*  **api:** add response comments PR ([62d03d2](https://github.com/AntaresSimulatorTeam/AntaREST/commit/62d03d2e7c5fe625dc5807236ea1807983624047))
*  **api:** fix tests ([214248f](https://github.com/AntaresSimulatorTeam/AntaREST/commit/214248f41be6acda7c7f626361a02b7e073374e2))
*  **api:** update yhe module service.py ([f1533d4](https://github.com/AntaresSimulatorTeam/AntaREST/commit/f1533d41a6a4b9bd8e22fe4d797f23fabf443d8a))
*  **api:** update the module raw_study_service.py ([2a857ab](https://github.com/AntaresSimulatorTeam/AntaREST/commit/2a857ab4a889cadfb86c256e2d0dce37ce8c8ba1))


### Contributors

<a href="https://github.com/MartinBelthle">MartinBelthle</a>, <a href="https://github.com/laurent-laporte-pro">laurent-laporte-pro</a>

<details>
<summary>Commit-Lint commands</summary>
<br />

You can trigger Commit-Lint actions by commenting on this PR:
- `@Commit-Lint merge patch` will merge dependabot PR on "patch" versions (X.X.Y - Y change)
- `@Commit-Lint merge minor` will merge dependabot PR on "minor" versions (X.Y.Y - Y change)
- `@Commit-Lint merge major` will merge dependabot PR on "major" versions (Y.Y.Y - Y change)
- `@Commit-Lint merge disable` will desactivate merge dependabot PR
- `@Commit-Lint review` will approve dependabot PR
- `@Commit-Lint stop review` will stop approve dependabot PR



</details>

laurent-laporte-pro committed:
test(api): simplify integration tests by using pytest fixtures

laurent-laporte-pro commented on antarest/study/web/studies_blueprint.py:
```diff
@@ -170,13 +170,26 @@ def copy_study(
+            current_user: Current authenticated user.
+        Returns:
+            The id of the copied study
+        """
```
It is important to use Markdown for a endpoint description. Also, you can improve the phrasing:

```markdown
This endpoint enables you to duplicate a study and place it in a specified location.
You can, for instance, copy a non-managed study to a managed workspace.

Args:
- `uuid`: The identifier of the study you wish to duplicate.
- `dest`: The destination workspace where the study will be copied.
- `with_outputs`: Indicates whether the study's outputs should also be duplicated. (Default value: False)
- `groups`: Specifies the groups to which your duplicated study will be assigned. (Default value: Current user's groups)
- `use_task`: Determines whether this duplication operation should trigger a task. It is recommended and set as the default value: True.
- `current_user`: Currently authenticated user.

Returns:
- The ID of the duplicated study.
```

laurent-laporte-pro commented on antarest/study/web/studies_blueprint.py:
```diff
@@ -170,13 +170,26 @@ def copy_study(
         )
         source_uuid = uuid
-        group_ids = groups.split(",") if groups else []
+        group_ids = groups.split(",") if groups else [group.id for group in current_user.groups]
```
You can consider using a `set` (or `frozenset`) to avoid duplicates.

laurent-laporte-pro commented on antarest/study/common/studystorage.py:
```diff
@@ -59,7 +59,7 @@ def exists(self, metadata: T) -> bool:
 
     @abstractmethod
-    def copy(self, src_meta: T, dest_name: str, with_outputs: bool = False) -> T:
+    def copy(self, src_meta: T, dest_name: str, groups: List[str], with_outputs: bool = False) -> T:
```
It is a better practice to use `Sequence[str]` instead of `List[str]`.

laurent-laporte-pro commented on antarest/study/storage/rawstudy/raw_study_service.py:
```diff
@@ -208,13 +208,15 @@ def copy(
         self,
         src_meta: RawStudy,
         dest_name: str,
+        groups: List[str],
```
It is a better practice to use `Sequence[str]` instead of `List[str]`.

laurent-laporte-pro commented on antarest/study/common/studystorage.py:
```diff
@@ -59,7 +59,7 @@ def exists(self, metadata: T) -> bool:
+    def copy(self, src_meta: T, dest_name: str, groups: List[str], with_outputs: bool = False) -> T:
         """
         Copy study to a new destination
         Args:
```
Could you correct the docstring?

laurent-laporte-pro commented on antarest/study/storage/rawstudy/raw_study_service.py:
```diff
@@ -208,13 +208,15 @@ def copy(
+        groups: List[str],
         with_outputs: bool = False,
     ) -> RawStudy:
         """
```
Could you correct/rephrase, for instance:

```python
"""
Copy Study to a New Destination

Args:
    src_meta: The source study.
    dest_name: The name of the destination study.
    groups: The groups to which the destination study will be assigned.
    with_outputs: Indicates whether the study outputs should be copied. (Specify 'True' to copy, 'False' to exclude.)

Returns:
    The destination study.
"""
```

laurent-laporte-pro commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -906,13 +906,15 @@ def copy(
         self,
         src_meta: VariantStudy,
         dst_name: str,
+        groups: List[str],
```
It is a better practice to use `Sequence[str]` instead of `List[str]`.

laurent-laporte-pro commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -906,13 +906,15 @@ def copy(
     ) -> VariantStudy:
         """
         Copy study to a new destination
         Args:
```
Again, could you rephrase?

laurent-laporte-pro commented on antarest/study/service.py:
```diff
@@ -813,7 +813,7 @@ def copy_study(
+        Copy study to another location.
 
         Args:
             src_uuid: source study
```
Documentation about the returned value is missing: It is a Task ID

laurent-laporte-pro commented on antarest/study/service.py:
```diff
@@ -1896,15 +1897,14 @@ def _save_study(
+        for gid in group_ids:
+            group = next(filter(lambda g: g.id == gid, owner.groups), None)
+            if group is None or not group.role.is_higher_or_equals(RoleType.WRITER) and not owner.is_site_admin():
+                raise UserHasNotPermissionError()
```
Add an error message. This is a "Permission denied" error.
You should also add an error message in the previous `raise UserHasNotPermissionError`.

laurent-laporte-pro commented on tests/study/storage/test_abstract_storage_service.py:
```diff
@@ -31,7 +31,7 @@ def exists(self, metadata: Study) -> bool:
 
     # noinspection SpellCheckingInspection
-    def copy(self, src_meta: Study, dest_name: str, with_outputs: bool = False) -> Study:
+    def copy(self, src_meta: Study, dest_name: str, groups: List[str], with_outputs: bool = False) -> Study:
```
It is a better practice to use `Sequence[str]` instead of `List[str]`.

laurent-laporte-pro reviewed (changes requested):
Typing and docstrings should be corrected.

Commit:
fix(api): response comments PR

Commit:
fix(api): add response comments PR

Commit:
fix(api): fix tests

Commit:
fix(api): update yhe module service.py

Commit:
fix(api): update the module raw_study_service.py

laurent-laporte-pro commented on antarest/study/service.py:
```diff
@@ -1891,20 +1892,19 @@ def _save_study(
         """
         if not owner:
-            raise UserHasNotPermissionError
+            raise UserHasNotPermissionError(f"{JWTUser} invalid for authentication")
```
Je pense qu'il faut changer ce message d'erreur, car il correspond à ceci :

```python
>>> from antarest.core.jwt import JWTUser
>>> print(f"{JWTUser} invalid for authentication")
<class 'antarest.core.jwt.JWTUser'> invalid for authentication
```

Ce genre de message est-il satisfaisant ?

laurent-laporte-pro committed:
feat(api): improve documentation and implementation of `copy_study` endpoint

- Updated the documentation for the `copy_study` endpoint to provide better clarity and details.
- Added parameter descriptions and explanations for the endpoint's purpose.
- Refactored the code for improved readability.
